### PR TITLE
Check required WordPress version to set galleryWithImageBlocks flag

### LINKED ISSUE #47782: Gallery block: Determine `galleryWithImageBlocks` prop value based on the WordPress version of the site when value is unknown
## Description

As described in https://github.com/wordpress-mobile/WordPress-Android/issues/17712, we pick the wrong version of the Gallery block when a self-hosted site is in WordPress 5.9+ and the Gutenberg plugin is not installed. This issue is mainly affecting the cases where the prop used to pick the Gallery version (`galleryWithImageBlocks`) is unknown, which leads to using `v1` since it's the default value. 

The Gallery block v2 is supported in WordPress 5.9+, so in order to avoid using the wrong version when having an unknown value, we could determine the Gallery block version to use based on the site's WordPress version.

I'm leaving some code references to pinpoint the logic used to determine the Gallery block version to use:

### `galleryWithImageBlocks` prop set by the app
iOS: https://github.com/WordPress/gutenberg/blob/705ab44f2c2797d75740533a85030c194b4fb040/packages/react-native-bridge/ios/Gutenberg.swift#L205-L207
Android: https://github.com/WordPress/gutenberg/blob/c840e90376da5de3e73af9499571e66fc8f25ff4/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/RNReactNativeGutenbergBridgeModule.java#L161-L164

### Picking Gallery block version based on `galleryWithImageBlocks`
https://github.com/WordPress/gutenberg/blob/70cf4fa090decbc841dadda5f7d2705c486839ce/packages/edit-post/src/editor.native.js#L29-L36
https://github.com/WordPress/gutenberg/blob/3356e6a12df5cfa3d2ed895af2a2be6fbfbc2a97/packages/block-library/src/gallery/shared.js#L36-L57
https://github.com/WordPress/gutenberg/blob/3356e6a12df5cfa3d2ed895af2a2be6fbfbc2a97/packages/block-library/src/gallery/edit-wrapper.js#L20-L24

## Step-by-step reproduction instructions

1. Create a self-hosted site with WordPress version 5.9 or above.
2. Check that the Gutenberg plugin is not installed. If not, uninstall it.
3. Create a post.
4. Insert a Gallery block.
5. Observe that the Gallery block is `v1` (content is not rendered using inner blocks).

## Expected behaviour

The correct version of the Gallery block should be used.

## Actual behaviour

In some cases, the wrong version of the Gallery block is used.

## Screenshots or screen recording (optional)

N/A

## WordPress information
- WordPress version: N/A
- Gutenberg version: N/A
- Are all plugins except Gutenberg deactivated? N/A
- Are you using a default theme (e.g. Twenty Twenty-One)? N/A

## Device information
- Device: N/A
- Operating system: N/A
- WordPress app version: `trunk`

### PR DESCRIPTION
**Pull request by derekblank**
Fixes https://github.com/WordPress/gutenberg/issues/47782:
* https://github.com/WordPress/gutenberg/issues/47782

The `galleryWithImageBlocks` param determines which version of the Gallery block should be used in the editor (v1 or v2). Gallery block v2 is supported in WordPress 5.9+ and should be used for all sites on WordPress 5.9 and above. This fix addresses an edge-case for self-hosted sites on WP 5.9+ that do not have the Gutenberg plugin installed. In this case, Gallery block v1 is the default, but Gallery block v2 should actually be used.

This code change checks the WordPress version and sets the param to true for any site using WP 5.9+, and keeps the original logic for sites not using WP 5.9+. Once WordPress 5.9 becomes the minimum supported version, this change can be removed.

An easy way to verify which version of the Gallery block is being used is to switch to HTML mode and check the tags in the markup:

Gallery block v2 syntax (`figure` + `image`):
```
<!-- wp:gallery {"linkTo":"none"} -->
<figure class="wp-block-gallery has-nested-images columns-default is-cropped">
    <!-- wp:image {"id":3326,"sizeSlug":"large","linkDestination":"none"} -->
    <figure class="wp-block-image size-large"><img
            src="https://****.files.wordpress.com/2022/05/image.jpg?w=1024" alt="" class="wp-image-3326" />
    </figure>
    <!-- /wp:image -->
</figure>
<!-- /wp:gallery -->
```

Gallery block v1 syntax (`figure` + `ul` + `li` + `image`):
```
<!-- wp:gallery {"ids":[38],"linkTo":"none"} -->
<figure class="wp-block-gallery columns-1 is-cropped">
    <ul class="blocks-gallery-grid">
        <li class="blocks-gallery-item">
            <figure><img src="http://localhost:8888/wp-content/uploads/2022/06/265-5000x5000-1-scaled.jpeg" data-id="38"
                    class="wp-image-38" /></figure>
        </li>
    </ul>
</figure>
<!-- /wp:gallery -->
```

To test:

1. Create a self-hosted site with WordPress version 5.9 or above.
2. Check that the Gutenberg plugin is not installed. If not, uninstall it.
3. Create a post.
4. Insert a Gallery block.
5. Observe that the Gallery block is using v2 (content is not rendered using inner blocks).

Bonus testing:
1. Create a self-hosted site with WordPress version 5.8 or below. (The [WP Downgrade](https://wordpress.org/plugins/wp-downgrade/) plugin can be used to quickly switch between WordPress versions.)
2. Repeat steps above and note that Gallery block v1 is used. 

Alternatively, when running locally, the value of `galleryWithImageBlocks` can be logged in the [editor](https://github.com/WordPress/gutenberg/blob/d3cf857e79b1c3359e94301b8a79564fb8ef0f85/packages/edit-post/src/editor.native.js#L36) to verify.

## Regression Notes
1. Potential unintended areas of impact
 - Block Editor Settings
 - Gutenberg Editor Settings
 - Experimental feature flags

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Determined manual testing and CI checks would provide an adequate amount of testing coverage for this edge case.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
N/A

### CODE REVIEW

File: WordPress/Classes/Models/BlockEditorSettings+GutenbergEditorSettings.swift [hunk 1]
```diff
@@ -12,7 +13,14 @@ extension BlockEditorSettings: GutenbergEditorSettings {
...
     public var galleryWithImageBlocks: Bool {
-        return experimentalFeature(.galleryWithImageBlocks)
+        // If site is using WP 5.9+ then return true as galleryWithImageBlocks is supported in WP 5.9+.
+        // Once support for WP 5.8 is dropped, this can be removed.
+        // https://github.com/WordPress/gutenberg/issues/47782
+        if hasRequiredWordPressVersion("5.9") {
+            return true
+        } else {
+            return experimentalFeature(.galleryWithImageBlocks)
+        }
```
Comment (derekblank) [inline WordPress/Classes/Models/BlockEditorSettings+GutenbergEditorSettings.swift]: In order to set the `hasRequiredWordPressVersion` flag and send it to the editor, we need access to the site's version of WordPress. This is handled by the Blog module in WPiOS, where the site's version is available as `self.version`. 

`hasRequiredWordPressVersion` is a method in [Blog.m](https://github.com/wordpress-mobile/WordPress-iOS/blob/5e1b64cdb4333ade0b272f84bc1ca534bdd75c3e/WordPress/Classes/Models/Blog.m#L953-L958) that checks that the installed WordPress version is more than or equal to the requiredVersion.

https://github.com/wordpress-mobile/WordPress-iOS/blob/5e1b64cdb4333ade0b272f84bc1ca534bdd75c3e/WordPress/Classes/Models/Blog.m#L953-L958

The idea is that could call this method from within [BlockEditorSettings+GutenbergEditorSettings.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/081d26f292bd2ee888a050306542c65270d3c185/WordPress/Classes/Models/BlockEditorSettings%2BGutenbergEditorSettings.swift) to detect the site's WordPress version.  Currently, the iOS build command will result in an error with the message "No such module Blog", but I would expect that this method would be available in Swift by importing the Blog module. I am less familiar with the way these modules should be shared and imported, however.

Essentially, we just need to use a piece of code from [Blog.m](https://github.com/wordpress-mobile/WordPress-iOS/blob/5e1b64cdb4333ade0b272f84bc1ca534bdd75c3e/WordPress/Classes/Models/Blog.m#L953-L958), and invoke it within [BlockEditorSettings+GutenbergEditorSettings.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/081d26f292bd2ee888a050306542c65270d3c185/WordPress/Classes/Models/BlockEditorSettings%2BGutenbergEditorSettings.swift).

@twstokes Any insight here on additional steps that would be needed to make this possible?

Comment (twstokes) [reply WordPress/Classes/Models/BlockEditorSettings+GutenbergEditorSettings.swift]: 👋 Hey @derekblank! `hasRequiredWordPressVersion` is a neat function, TIL. Thanks for taking the time to get me up to speed on this. 🙇 

The good news is that the method `hasRequiredWordPressVersion` should already be exposed to this file and we just need to call it on an existing `Blog` object, so I don't think we'll need to deal with any bridging or exposing.

What may be even better news is that it looks like a reference to the parent blog [is already available to the class](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Models/BlockEditorSettings%2BCoreDataProperties.swift#L38). If this works like I hope, you may be able to call `blog.hasRequiredWordPressVersion("5.9")` here and remove the `Blog` import. 

Let me know if that helps!

Comment (derekblank) [reply WordPress/Classes/Models/BlockEditorSettings+GutenbergEditorSettings.swift]: > What may be even better news is that it looks like a reference to the parent blog [is already available to the class](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Models/BlockEditorSettings%2BCoreDataProperties.swift#L38). If this works like I hope, you may be able to call blog.hasRequiredWordPressVersion("5.9") here and remove the Blog import.

That's a bingo. After making this change, the app builds, and the fix is successful in my testing, so I think we can mark this PR as ready for review. Thanks for taking a look. 👀

**Review by twstokes — COMMENTED** [review]
Thanks Derek for tackling this PR! 🚀 

### Results of my testing on a self-hosted site:

| 5.8 - `trunk` | 5.9 - `trunk` | 5.8 - this PR | 5.9 - this PR |
| - | - | - | - |
| Simulator Screenshot - iPhone 14 Pro - 2023-05-25 at 23 21 42 | Simulator Screenshot - iPhone 14 Pro - 2023-05-25 at 22 49 37 | Simulator Screenshot - iPhone 14 Pro - 2023-05-25 at 22 59 00 | Simulator Screenshot - iPhone 14 Pro - 2023-05-25 at 22 57 04 |

From my understanding of the difference between `v1` and `v2` Gallery blocks ([thanks to this PR](https://github.com/WordPress/gutenberg/pull/41533)), this change appears to pass the tests in this PR. ✅ 

### Comments / observations

- We should add an internal release note
- A `bug` label should be added as well as `Posting/Editing`
- A milestone should be added
- Consider adding expected output (e.g. what is shown in HTML mode). Readers outside of the Gutenberg world may not easily understand what inner blocks are or other differences between `v1` and `v2` (I struggled myself)
- The four items in the PR submission checklist should be checked before getting reviews
- Consider a note about what research you did regarding tests (since one isn't added)

Thanks!

File: RELEASE-NOTES.txt [hunk 2]
```diff
@@ -1,3 +1,7 @@
+22.6
+----
+* [**] Check required WordPress version to set galleryWithImageBlocks flag [#20736]
```
Comment (fluiddot) [inline RELEASE-NOTES.txt]: As mentioned [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/20736#pullrequestreview-1444921355), we should add the `[internal]` prefix here.

Comment (fluiddot) [reply RELEASE-NOTES.txt]: Nitpick:
```suggestion
* [**] Check required WordPress version to set "galleryWithImageBlocks" flag [#20736]
```

**Review by fluiddot — COMMENTED** [review]
I've tested different scenarios and confirmed the change works as expected 🎊 .

- WP 5.8.7 + Gutenberg plugin disabled => Gallery v1 ✅
- WP 5.8.7 + Gutenberg plugin 13.6.0 => Gallery v1 ✅
- WP 5.9 + Gutenberg plugin disabled => Gallery v2 **(test case that this PR aims to fix)** ✅
- WP 5.9 + Gutenberg plugin 13.6.0 => Gallery v2 ✅

**NOTE:** The latest version of the Gutenberg plugin that supports WP 5.8 is version 13.6.0 (The minimum supported WP version was set to 5.9 in Gutenberg plugin version 13.7).

File: RELEASE-NOTES.txt (hunk 2 above)
Comment (derekblank) [reply RELEASE-NOTES.txt]: I made both changes. 👍 Out of curiosity, and so I can follow convention in the future, was the quotation marks suggestion simply to note that `galleryWithImageBlocks` is code, or is there perhaps another purpose of the quotation marks?

Comment (fluiddot) [reply RELEASE-NOTES.txt]: Yeah, the nitpick suggestion of using quotation marks is mainly to denote `galleryWithImageBlocks` is code.

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! I re-tested what I did in [the last review](https://github.com/wordpress-mobile/WordPress-iOS/pull/20736#pullrequestreview-1446156903) and confirmed this change works as expected ✅.
 
However, as I mentioned [here](https://github.com/wordpress-mobile/WordPress-Android/issues/17712#issuecomment-1566972693), seems we still have an issue when using Gallery v1 and setting a caption that would be great to investigate.

**Comment by derekblank** [comment]
Thanks for reviewing @fluiddot! I've addressed the Caption issue separately here:

* https://github.com/WordPress/gutenberg/pull/51076
